src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1050,6 +1059,67 @@ public void flushAndClose() throws IOException {
         }
     }
 
+    protected TranslogRecoveryStats recoverFromTranslog(long translogId, TranslogRecoveryPerformer handler) throws IOException {
+        final Translog translog = engineConfig.getTranslog();
+        Set<String> recoveredTypes = new HashSet<>();
+        int operationsRecovered = 0;
+        boolean failed = false;
+        try (Translog.OperationIterator in = translog.openIterator(translogId)) {
+            Translog.Operation operation;
+            while ((operation = in.next()) != null) {
+                try {
+                    final IndexingOperation indexingOperation = handler.performRecoveryOperation(this, operation);
+                    if (indexingOperation != null && indexingOperation.parsedDoc().mappingsModified()) {
+                        recoveredTypes.add(indexingOperation.type());
+                    }
+                    operationsRecovered++;
+                } catch (ElasticsearchException e) {
+                    if (e.status() == RestStatus.BAD_REQUEST) {
+                        // mainly for MapperParsingException and Failure to detect xcontent
+                        logger.info("ignoring recovery of a corrupt translog entry", e);
+                    } else {
+                        throw e;
+                    }
+                }
+            }
+        } catch (FileNotFoundException ex) {
+            logger.info("no translog file found for ID: " + translogId);
+        } catch (TruncatedTranslogException e) {
+            // file is empty or header has been half-written and should be ignored
+            logger.trace("ignoring truncation exception, the translog is either empty or half-written", e);
+        } catch (Throwable e) {
+            failed = true;
+            IOUtils.closeWhileHandlingException(translog);
+            throw new EngineException(shardId, "failed to recover from translog", e);
+        } finally {
+            if (failed == false) {
```

Comment:
Is the finally and failed flag really needed? Could you move the flush code outside the try/catch, in which case it would only be executed in the case where we do not throw EngineException above (which is what the current code does, but with using this flag)?